https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1974/ shows this error,
```
[pytorch-container] The command '/bin/sh -c pip install -q torch==1.0.* torchvision==0.2.*' returned a non-zero code: 1 
CI_build.Makefile:472: recipe for target 'pytorch-container' failed
make: *** [pytorch-container] Error 1
```
Temporary network issue?

--- a/clipper_admin/clipper_admin/docker/docker_container_manager.py
+++ b/clipper_admin/clipper_admin/docker/docker_container_manager.py
@@ -23,6 +23,7 @@ from .logging.docker_logging_utils import (
     get_default_log_config
 )
 from clipper_admin.docker.logging.fluentd import Fluentd
+from clipper_admin.decorators import retry
 
 logger = logging.getLogger(__name__)
 
@@ -181,6 +182,7 @@ class DockerContainerManager(ContainerManager):
         # Redis for cluster configuration
         if not self.external_redis:
             self.logger.info("Starting managed Redis instance in Docker")
+            redis_name = "redis-{}".format(random.randint(0, 100000))
             self.redis_port = find_unbound_port(self.redis_port)
             redis_labels = self.common_labels.copy()
             redis_labels[CLIPPER_DOCKER_PORT_LABELS['redis']] = str(
@@ -189,18 +191,19 @@ class DockerContainerManager(ContainerManager):
                 'redis:alpine',
                 "redis-server --port %s" % CLIPPER_INTERNAL_REDIS_PORT,
                 log_config=self.log_config,
-                name="redis-{}".format(random.randint(
-                    0, 100000)),  # generate a random name
+                name=redis_name,
                 ports={
                     '%s/tcp' % CLIPPER_INTERNAL_REDIS_PORT: self.redis_port
                 },
                 labels=redis_labels,
                 **self.extra_container_kwargs)
             self.redis_ip = redis_container.name
+            self._check_container_status(redis_name)
 
         # frontend management
         mgmt_cmd = "--redis_ip={redis_ip} --redis_port={redis_port}".format(
             redis_ip=self.redis_ip, redis_port=CLIPPER_INTERNAL_REDIS_PORT)
+        mgmt_name = "mgmt_frontend-{}".format(random.randint(0, 100000))
         self.clipper_management_port = find_unbound_port(
             self.clipper_management_port)
         mgmt_labels = self.common_labels.copy()
@@ -211,14 +214,14 @@ class DockerContainerManager(ContainerManager):
             mgmt_frontend_image,
             mgmt_cmd,
             log_config=self.log_config,
-            name="mgmt_frontend-{}".format(random.randint(
-                0, 100000)),  # generate a random name
+            name=mgmt_name,
             ports={
                 '%s/tcp' % CLIPPER_INTERNAL_MANAGEMENT_PORT:
                 self.clipper_management_port
             },
             labels=mgmt_labels,
             **self.extra_container_kwargs)
+        self._check_container_status(mgmt_name)
 
         # query frontend
         query_cmd = ("--redis_ip={redis_ip} --redis_port={redis_port} "
@@ -254,6 +257,7 @@ class DockerContainerManager(ContainerManager):
             },
             labels=query_labels,
             **self.extra_container_kwargs)
+        self._check_container_status(query_name)
 
         # Metric Section
         query_frontend_metric_name = "query_frontend_exporter-{}".format(
@@ -262,6 +266,7 @@ class DockerContainerManager(ContainerManager):
             query_frontend_metric_name, self.docker_client, query_name,
             frontend_exporter_image, self.common_labels,
             self.log_config, self.extra_container_kwargs)
+        self._check_container_status(query_frontend_metric_name)
 
         self.prom_config_path = tempfile.NamedTemporaryFile(
             'w', suffix='.yml', delete=False).name
@@ -272,14 +277,18 @@ class DockerContainerManager(ContainerManager):
         setup_metric_config(query_frontend_metric_name, self.prom_config_path,
                             CLIPPER_INTERNAL_METRIC_PORT)
 
+        metric_frontend_name = "metric_frontend-{}".format(
+            random.randint(0, 100000))
         self.prometheus_port = find_unbound_port(self.prometheus_port)
         metric_labels = self.common_labels.copy()
         metric_labels[CLIPPER_DOCKER_PORT_LABELS['metric']] = str(
             self.prometheus_port)
         metric_labels[CLIPPER_METRIC_CONFIG_LABEL] = self.prom_config_path
-        run_metric_image(self.docker_client, metric_labels,
-                         self.prometheus_port, self.prom_config_path,
-                         self.log_config, self.extra_container_kwargs)
+        run_metric_image(metric_frontend_name, self.docker_client,
+                         metric_labels, self.prometheus_port,
+                         self.prom_config_path, self.log_config,
+                         self.extra_container_kwargs)
+        self._check_container_status(metric_frontend_name)
 
         self.connect()
 
@@ -404,15 +413,7 @@ class DockerContainerManager(ContainerManager):
                                                    image)
                 model_container_names.append(container_name)
             for name in model_container_names:
-                container = self.docker_client.containers.get(name)
-                while True:
-                    state = container.attrs.get("State")
-                    inspect_container = self.docker_client.api.inspect_container(name)
-                    if (state is not None and state.get("Status") == "running") or \
-                            (inspect_container is not None and
-                             inspect_container.get("State").get("Health").get("Status") == "healthy"):
-                        break
-                    time.sleep(3)
+                self._check_container_status(name)
 
         elif len(current_replicas) > num_replicas:
             num_extra = len(current_replicas) - num_replicas
@@ -474,6 +475,19 @@ class DockerContainerManager(ContainerManager):
                 c.stop()
             else:
                 c.kill()
+
+    # Wait for maximum 5 min.
+    @retry((docker.errors.NotFound, docker.errors.APIError, ClipperException),
+           tries=300, delay=1, backoff=1, logger=logger)
+    def _check_container_status(self, name):
+        state = self.docker_client.containers.get(name).attrs.get("State")
+        inspected = self.docker_client.api.inspect_container(name)
+        if (state is not None and state.get("Status") == "running") or \
+           (inspected is not None and inspected.get("State").get("Health").get("Status") == "healthy"):
+            return
+        else:
+            msg = "{} container is not running yet".format(name)
+            raise ClipperException(msg)
 
     def _is_valid_logging_state_to_connect(self, all_labels):
         if self.centralize_log and not self.logging_system.container_is_running(all_labels):

--- a/clipper_admin/clipper_admin/docker/docker_metric_utils.py
+++ b/clipper_admin/clipper_admin/docker/docker_metric_utils.py
@@ -72,10 +72,12 @@ def setup_metric_config(query_frontend_metric_name, prom_config_path,
         yaml.dump(prom_config, f)
 
 
-def run_metric_image(docker_client, common_labels, prometheus_port,
-                     prom_config_path, log_config, extra_container_kwargs):
+def run_metric_image(metric_frontend_name, docker_client, common_labels,
+                     prometheus_port, prom_config_path, log_config,
+                     extra_container_kwargs):
     """
     Run the prometheus image.
+    :param metric_frontend_name: container name
     :param docker_client: The docker client object
     :param common_labels: Labels to pass in
     :param prom_config_path: Where config file lives
@@ -95,7 +97,7 @@ def run_metric_image(docker_client, common_labels, prometheus_port,
     docker_client.containers.run(
         "prom/prometheus:{}".format(PROM_VERSION),
         metric_cmd,
-        name="metric_frontend-{}".format(random.randint(0, 100000)),
+        name=metric_frontend_name,
         ports={'9090/tcp': prometheus_port},
         log_config=log_config,
         volumes={

--- a/integration-tests/clipper_metric_docker.py
+++ b/integration-tests/clipper_metric_docker.py
@@ -104,7 +104,7 @@ if __name__ == '__main__':
         cleanup=False, start_clipper=True, new_name=cluster_name)
     python_deployer.create_endpoint(
         clipper_conn, "simple-example", "doubles", feature_sum, num_replicas=2)
-    time.sleep(2)
+    time.sleep(10)
     try:
         logger.info(
             "Making 100 predictions using two model container; Should takes 25 seconds."

--- a/integration-tests/clipper_metric_docker.py
+++ b/integration-tests/clipper_metric_docker.py
@@ -3,13 +3,13 @@ from __future__ import print_function
 import json
 import logging
 import os
-import subprocess
 import sys
 import time
 import random
 
 import numpy as np
 import requests
+from requests.exceptions import RequestException
 import yaml
 from test_utils import log_clipper_state, create_docker_connection
 
@@ -17,6 +17,16 @@ cur_dir = os.path.dirname(os.path.abspath(__file__))
 sys.path.insert(0, os.path.abspath("%s/../clipper_admin" % cur_dir))
 from clipper_admin import ClipperConnection, DockerContainerManager
 from clipper_admin.deployers import python as python_deployer
+from clipper_admin.decorators import retry
+from clipper_admin.exceptions import ClipperException
+
+logging.basicConfig(
+    format=
+    '%(asctime)s %(levelname)-8s [%(filename)s:%(lineno)d] %(message)s',
+    datefmt='%y-%m-%d:%H:%M:%S',
+    level=0)
+
+logger = logging.getLogger(__name__)
 
 
 def predict(addr, x):
@@ -38,12 +48,16 @@ def get_metrics_config():
     return conf
 
 
+# Wait for maximum 5 min.
+@retry((RequestException, ClipperException),
+       tries=300, delay=1, backoff=1, logger=logger)
 def get_matched_query(metric_addr, metric_name):
     query = gen_match_query(metric_addr, metric_name)
     logger.info("Querying: {}".format(query))
-    print('test why docker query is broken: {}'.format(repr(query)))
     res = requests.get(query).json()
     logger.info(res)
+    if len(res['data']) == 0:
+        raise ClipperException("returned data is empty")
     return res
 
 
@@ -52,11 +66,16 @@ def parse_res_and_assert_node(res, node_num):
     assert len(res['data']) == node_num
 
 
+# Wait for maximum 5 min.
+@retry((RequestException, ClipperException),
+       tries=300, delay=1, backoff=1, logger=logger)
 def check_target_health(metric_addr):
     query = metric_addr + '/api/v1/targets'
     logger.info("Querying: {}".format(query))
     res = requests.get(query).json()
     logger.info(res)
+    if len(res['data']) == 0:
+        raise ClipperException("returned data is empty")
     assert res['status'] == 'success'
 
     active_targets = res['data']['activeTargets']
@@ -77,14 +96,6 @@ def log_docker_ps(clipper_conn):
 if __name__ == '__main__':
     # metric_addr = "http://localhost:9090"
     gen_match_query = lambda addr, name: "http://{addr}/api/v1/series?match[]={name}".format(addr=addr, name=name)
-
-    logging.basicConfig(
-        format=
-        '%(asctime)s %(levelname)-8s [%(filename)s:%(lineno)d] %(message)s',
-        datefmt='%y-%m-%d:%H:%M:%S',
-        level=0)
-
-    logger = logging.getLogger(__name__)
 
     logger.info("Start Metric Test (0/1): Running 2 Replicas")
 

